src/fsharp/Optimizer.fs

```diff
@@ -1763,16 +1763,23 @@ let TryDetectQueryQuoteAndRun cenv (expr:Expr) =
     | _ -> 
         //printfn "Not eliminating because no Run found"
         None
+
+let IsILMethodRefDeclaringTypeSystemString (ilg: ILGlobals) (mref: ILMethodRef) =
+    mref.DeclaringTypeRef.Scope.IsAssemblyRef &&
+    mref.DeclaringTypeRef.Scope.AssemblyRef.Name = ilg.typ_String.TypeRef.Scope.AssemblyRef.Name &&
```

Comment:
@TIHan 
There are a few more parts to the identity than just the AssemblyName,  do you think we might need to compare the public key, also:

Also references to strings can be to any of the the assemblies mscorlib, System.Runtime, and System.Private.CoreLib.  are we sure they have all been unified at this point.  Or should we just special case the comparison, to ensure it matches one of these?

````
  type AssemblyRef = 
    { Name: string;
      Hash: Option<Bytes>;
      PublicKeyInfo: Option<PublicKeyInfo>;
      Version: Option<VersionInfo>; 
      Locale: Option<Locale>; } 
````

Reply:
I don't think we need to look at the public key. The assembly name should be sufficient for this type of optimization. 

We do this in other places in the optimizer already. If you look at https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/Optimizer.fs#L1983 , it only looks as the assembly name.

Reply:
To answer your question on if we can guarantee that it comes from assemblies mscorlib, System.Runtime, System.Private.CoreLib - We can as the construction of `ILGlobals` gets passed in a scope with an assembly ref of either mscorlib, System.Runtime, etc. https://github.com/Microsoft/visualfsharp/blob/e2d83009a83bc92c6518c2b4409f1e1a1417ce13/src/absil/il.fsi#L1565

Reply:
@TIHan 

The example you pointed to **suggests** that the optimizer doesn't optimize between references to System.Runtime and mscorlib for Array.Length unless the ref has been normalized.

I.e.  if the DeclaringTypeRef is referencing System.Runtime, and the compiler global typ_Array points to the System.Private.CoreLib implementation.  So that optimization code is skipped.

I **suspect** the same would be true after you implement this additional check.

The impact of this check is basically less optimization in scenarios we think should work, in support of not being broken in the scenario where a developer references a type in an assembly that has the same name as a system type.

As for private keys, the array code and this string code is broken, in the circumstance where we expect a user to provide a system type in a system assembly that merely uses a different key. Personally I think any developer who does that is crazy, it is a wierd scenario.

Reply:
> if the DeclaringTypeRef is referencing System.Runtime, and the compiler global typ_Array points to the System.Private.CoreLib implementation. So that optimization code is skipped.

I don't see this happening. If `System.Runtime` scope ref was passed to `ILGlobals`, then any use of string `ILType` will refer to `System.Runtime` as its assembly ref. If `System.Private.CoreLib` scope ref was passed to `ILGlobals`, same thing. All string `ILType`s will be imported by whatever `ILGlobals` `typ_String` is - see here: https://github.com/Microsoft/visualfsharp/blob/master/src/absil/ilread.fs#L2049
I would argue that if it is possible for this to happen, to have two string `ILType`'s with different assembly refs when typechecking/optimizing/etc, then that is a bug. But, I don't think that is happening.

> As for private keys, the array code and this string code is broken, in the circumstance where we expect a user to provide a system type in a system assembly that merely uses a different key. Personally I think any developer who does that is crazy, it is a wierd scenario.

We won't hit this situation because of the same reason I mentioned above regarding `ILGlobals` and scope refs. They will always have the same key.

Perhaps we should discuss if we should be checking assemblies in general. Maybe the 'shape' is all we care about and not the assembly.